### PR TITLE
Cast half/bfloat to single before neg. 

### LINKED ISSUE #1541: Simple slice fusion fails to segment
Minimal repro:
```python
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1], contiguity=[True], dtype=DataType.BFloat16)
    T1 = fd.ops.neg(T0)
    T2 = fd.ops.slice(T1, start_indices=[0], end_indices=[5])
    fd.add_output(T2)

with FusionDefinition() as fd:
    nvfuser_fusion_id1(fd)

inputs = [
    torch.randn((10,), dtype=torch.bfloat16, device='cuda:0')
]
fd.execute(inputs)
```
This fails with `Can not find a scheduler to schedule fusion segment`. `NVFUSER_DUMP=segmenter_logging` shows:
```
***Runtime***: Try to schedule fusion un-segmented:

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _pointwise_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _inner_persistent_ ***rejected*** because : needs a reduction op
Scheduler _outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _inner_outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _matmul_ ***rejected*** because : Matmul scheduler supports fusions only with a single MMA op, got: 0

**Segmenter** Considering fusion:
i13 = fmin(i0, 5);
i15 = fmax(0, i13);
T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
   = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
***Accepted*** as: pointwise

**Segmenter** Considering fusion:
T1_g[ iS1{i0} ]
   = -T0_g[ iS0{i0} ];
i13 = fmin(i0, 5);
i15 = fmax(0, i13);
T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
   = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _pointwise_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _inner_persistent_ ***rejected*** because : needs a reduction op
Scheduler _outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _inner_outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _matmul_ ***rejected*** because : Matmul scheduler supports fusions only with a single MMA op, got: 0
An error occurred while executing nvFuser FusionDefinition 1.
```

### PR DESCRIPTION
**Pull request by wujingyue**
Fixes #1541.

### CODE REVIEW

File: csrc/type.cpp
Comment (wujingyue) [inline csrc/type.cpp:1]: Changes to this file are not necessary for the fix. They are part of my failed attempt to support unsigned types for neg. I could revert them but think they are good additions anyway. Let me know if you hate them.

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
!build

File: test/test_unary.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    // Skip unsigned types because they are uncommon and not supported in many
+    // places, e.g.,
+    // https://github.com/NVIDIA/Fuser/blob/329c3b194e7de5bccd3122468893c1fa83cd2a2e/csrc/executor.cpp#L478.
+    testing::Values(
+        PrimDataType::Half,
+        PrimDataType::BFloat16,
+        PrimDataType::Float,
+        PrimDataType::Double,
+        PrimDataType::Int32,
+        PrimDataType::Int),
```
Comment (jacobhinkle) [inline test/test_unary.cpp:81]: Nice thorough test, and I like the use of the gtest test parametrization here. Is there any need to test complex types here?

File: csrc/type_promotion.h [hunk 2]
```diff
@@ -13,20 +13,9 @@
...
-//!
-//! 1) check_all_same_dtype_ flag checks that all inputs and defined outputs
-//! have the same dtype. Default = False
-//!
-//! 2) promote_inputs_to_common_dtype flag will cast the inputs to the common
-//! dtype. Default = True
-//!
-//! 3) promote_integer_inputs_to_float flag will cast the common dtype to the
-//! default float scalar type if it is an integral type (including bool).
-//!
```
Comment (jacobhinkle) [inline csrc/type_promotion.h:27]: Was this comment removed because it's out of date? Maybe we can just update it instead.

**Review by jacobhinkle — APPROVED** [review]
LGTM. Just a couple small comments. Thanks for the fix!

File: test/test_unary.cpp (hunk 1 above)
Comment (wujingyue) [reply test/test_unary.cpp:81]: Done.

File: csrc/type_promotion.h (hunk 2 above)
Comment (wujingyue) [reply csrc/type_promotion.h:27]: > Was this comment removed because it's out of date?

Yes. 

> Maybe we can just update it instead.

It's no longer fair to say the flags here are derived from ATen/TensorIterator.h. 
promote_integer_inputs_to_float comes from ATen/TensorIterator.h and require_full_precision_promoted doesn't. We also miss [a dozen of flags there](https://github.com/pytorch/pytorch/blob/242e03ba8661a62a1cf54ac89bda555b2120a6d4/aten/src/ATen/TensorIterator.h#L951-L961) if our intention is to match.

**Comment by wujingyue** [comment]
!build

File: test/test_unary.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      NVF_ERROR(
+          isFloatingPointType(dtype) || isComplexType(dtype),
+          "Expect a floating point type or a complex type, but found: ",
+          dtype);
+      in_tensor = at::randn(shape, options);
+  }
+
+  FusionExecutorCache fec(std::move(fusion));
+  auto out_tensors = fec.runFusionWithInputs({in_tensor});
+  // Calculate the reference output explicitly. Type promotion happens when
```
Comment (jjsjann123) [inline test/test_unary.cpp:62]: :+1:

File: csrc/ops/arith.cpp [hunk 4]
```diff
@@ -447,14 +447,28 @@ TensorView* eye(Val* size, DataType dtype) {
...
 NVFUSER_DEFINE_UNARY_OP(print, Print)
 #undef NVFUSER_DEFINE_UNARY_OP
 
+// Promote half to single for neg. Eventually, `neg` should probably be defined
+// using NVFUSER_DEFINE_UNARY_OP. However, currently, nvFuser codegen misses
+// certain header files for half types and therefore has no access to data types
+// like `__nv_bfloat16`  and intrinsics like `__hneg`. So, as a workaround to
+// #1541, we promote half types to single.
+//
+// Caveat: for `neg`, don't promote integer types to single.
```
Comment (jjsjann123) [inline csrc/ops/arith.cpp:463]: nitpick: this reads like we *are* promoting integer to single... but we are not, right?!

**Review by jjsjann123 — COMMENTED** [review]
LGTM. Thanks for the nice comment in `arith.cpp`

**Review by jjsjann123 — APPROVED** [review]
meant to stamp.

File: csrc/ops/arith.cpp (hunk 4 above)
Comment (wujingyue) [reply csrc/ops/arith.cpp]: I rewrote the comment. I hope it's more clear now.
